src/event/event-to-string.cc

```diff
@@ -0,0 +1,127 @@
+#include "bindings.h"
+
+//
+// Convert an event's metadata to a string representation.
+//
+Napi::Value Event::toString(const Napi::CallbackInfo& info) {
+  Napi::Env env = info.Env();
+  char buf[Event::fmtBufferSize];
+
+  int rc;
+  // no args is non-human-readable form - no delimiters, uppercase
+  // if arg == 1 it's the original human readable form.
+  // otherwise arg is a bitmask of options.
+  if (info.Length() == 0) {
+    rc = oboe_metadata_tostr(&this->event.metadata, buf, sizeof(buf) - 1);
+  } else {
+    int style = info[0].ToNumber().Int64Value();
+    int flags;
+    // make style 1 the previous default because ff_header alone is not very
+    // useful.
+    if (style == 1) {
+      flags = Event::ff_header | Event::ff_task | Event::ff_op |
+              Event::ff_flags | Event::ff_separators | Event::ff_lowercase;
+    } else {
+      // the style are the flags and the separator is a dash.
+      flags = style;
+    }
+    rc = format(&this->event.metadata, sizeof(buf), buf, flags) ? 0 : -1;
+  }
+
+  return Napi::String::New(env, rc == 0 ? buf : "");
+}
+
+//
+// core formatting code
+//
+//
+// function to format an x-trace with components split by sep.
+//
+int Event::format(oboe_metadata_t* md, size_t len, char* buffer, uint flags) {
```

Comment:
would it make sense to check for NULL pointers in `md` and `buffer` to protect against segfaults?

Reply:
i don't think it can happen but maybe you can correct my thinking. my logic:
- the function is only called from `Event::toString()` immediately above
- buf is allocated on the stack in `toString()` and cannot be null
- metadata is `&this->event.metadata` so only if `this` is invalid can metadata be a null pointer.

if `this` is messed up it can be null or a random value. the only case in which `md` will be null is if `this` was null (`event` is offset zero in the `Event` instance (i think) and `metadata` is offset 0 of `event`). but before that the code `&this->event.metadata` in `toString()` is referencing `this` as a pointer so it seems like it would segfault there.

unless i don't trust `this` i can't see how this can happen.

Reply:
Yes, it probably works the way this function is currently called. It's usually good though not having to depend on the caller for valid data as this function could be called from elsewhere in the future. Just another level of safety, but that's up to you. Just a thought.

Reply:
in general when something is internal-only i am willing to count on good data. but you make a good point - nothing is restricting using this function outside of this file, so i have changed it to where format() is not a member of the Event class and can only be called from this single file.

Reply:
:+1: